# Monkeybread: Analyze single-cell spatial transcriptomics data

![PyPI Version](https://img.shields.io/pypi/v/monkeybread)
![Read the Docs](https://readthedocs.org/projects/monkeybread/badge/?version=latest)

## About

Monkeybread is a Python package that facilitates the analysis of single-cell resolution spatial transcriptomics data such as those generated by the [MERSCOPE](https://vizgen.com/?gclid=CjwKCAjwivemBhBhEiwAJxNWN_NME_aGNx1gyfe2ToEYrY3LT35kHPrd5g042aIZVTsUqz3kdgXmQhoCWn4QAvD_BwE) or [Xenium](https://www.10xgenomics.com/platforms/xenium) platforms.

Monkeybread provides tools that enable:
* Identification of cellular niches (i.e., regions with distinct compositions of cell types)
* Visualization of density of cell types across the tissue
* Statistical tests for testing for colocalization between cell types
* Statistical tests and visualiziation for ligand-receptor co-expression between neighboring cells

Monkeybread operates on datasets stored as [AnnData](https://anndata.readthedocs.io/en/latest/) objects and thus, can be integrated into analyses that use packages from the [scverse](https://scverse.org/) such as [scanpy](https://scanpy.readthedocs.io/en/latest/) or [squidpy](https://squidpy.readthedocs.io/en/stable/).

Monkeybread was developed at [Immunitas Therapeutics](https://www.immunitastx.com/). 

## Installation

To install Monkeybread using pip, run the following command:

`pip install monkeybread`

## Usage

### Documentation

For Monkeybread's API manual, please visit the [documentation](https://monkeybread.readthedocs.io/en/latest/index.html).

### Tutorial

For a tutorial on running Monkeybread, please see the [tutorial](https://monkeybread.readthedocs.io/en/latest/notebooks/tutorial.html).
